app/models/event.rb

```diff
@@ -610,4 +614,63 @@ def fix_keywords
   def presence_default
     self.presence = :onsite if presence.blank?
   end
+
+  def set_status_and_notify
+    if user.has_role?('trusted_user') || user.has_role?('admin')
+      # Check if the user creating the event has a 'trusted' role or is admin.
+      # If true, mark the event status as 'approved' and send a notification email to the user.
+      self.update!(event_status: Event.event_statuses[:approved])
+      UserMailer.event_published(self).deliver_later if user.has_role?('trusted_user')
+    else
+      # If the user is not trusted, the event requires admin review.
+      # Send a notification email to the admin to review the event.
+
+      # deliver_later is asynchronous. When you use this method, the email is not send at the moment,
+      # but rather is pushed in a job's queue. If the job is not running, the email will not be sent.
+      # Deliver_now will send the email at the moment, no matter what is the job's state.
+      AdminMailer.review_event(self).deliver_later
+    end
+  end
+
+
+  def change_status_and_notify
+    if self.previous_changes.key?("event_status")
+      # getting event status change i.e. old_status and new_status
+      old_status, new_status = self.previous_changes["event_status"]
+
+      # getting the status constants from enum, rather then hardcoding
+      awaiting_review = Event.event_statuses.key(0)
+      approved = Event.event_statuses.key(1)
+      revisions_required = Event.event_statuses.key(3)
+
+      # Check if the event status has changed to 'approved'
+      if (old_status == awaiting_review && new_status == approved) ||
+        (old_status == revisions_required && new_status == approved)
+
+        # Increment and save the approved events count
+        self.user.update!(approved_events_count: self.user.approved_events_count + 1)
+
+        # Fetch role information
+        trusted_user_role = Role.find_by(title: "Trusted user")
+        registered_user_role = Role.find_by(title: "Registered user")
+
+        # Ensure both roles exist
```

Comment:
if you are asking that why do we need to ensure that the roles are there, its just a good safety check. so that even if the trusted user role is not there, the code would not crash or something

Reply:
but by that logic we should be checking for existence of every role before every call involving a role?

Reply:
We don’t do this for every role because most role checks are used for access control (not core logic) and usually have fallback behavior if a role is missing. 

However, in this method, the roles are central to the business logic (i.e., upgrading the user’s role), so it’s important to verify their existence to prevent unexpected issues.